src/libcollections/slice.rs

```diff
@@ -1305,213 +1304,290 @@ impl<T: Clone> ToOwned for [T] {
-        LARGE_INSERTION
-    };
+        // The right run is shorter.
+        ptr::copy_nonoverlapping(v_mid, buf, len - mid);
+        guard = Guard {
+            start: buf,
+            end: buf.offset((len - mid) as isize),
+            dst: v_mid,
+        };
+
+        let left = &mut guard.dst;
+        let right = &mut guard.end;
+        let mut out = v_end;
+
+        while v < *left && buf < *right {
+            // Consume the greater side.
+            // If equal, prefer the right run to maintain stability.
+            let to_copy = if compare(&*left.offset(-1), &*right.offset(-1)) == Greater {
+                backward(left)
+            } else {
+                backward(right)
+            };
+            ptr::copy_nonoverlapping(to_copy, backward(&mut out), 1);
+        }
+    }
 
-    let len = v.len();
+    // Increments the pointer and returns the old value.
+    unsafe fn forward<T>(ptr: &mut *mut T) -> *mut T {
+        let old = *ptr;
+        *ptr = ptr.offset(1);
+        old
+    }
 
-    // short vectors get sorted in-place via insertion sort to avoid allocations
-    if len <= insertion {
-        insertion_sort(v, compare);
-        return;
+    // Decrements the pointer and returns the new value.
+    unsafe fn backward<T>(ptr: &mut *mut T) -> *mut T {
+        *ptr = ptr.offset(-1);
+        *ptr
     }
 
-    // allocate some memory to use as scratch memory, we keep the
-    // length 0 so we can keep shallow copies of the contents of `v`
-    // without risking the dtors running on an object twice if
-    // `compare` panics.
-    let mut working_space = Vec::with_capacity(2 * len);
-    // these both are buffers of length `len`.
-    let mut buf_dat = working_space.as_mut_ptr();
-    let mut buf_tmp = unsafe { buf_dat.offset(len as isize) };
-
-    // length `len`.
-    let buf_v = v.as_ptr();
-
-    // step 1. sort short runs with insertion sort. This takes the
-    // values from `v` and sorts them into `buf_dat`, leaving that
-    // with sorted runs of length INSERTION.
-
-    // We could hardcode the sorting comparisons here, and we could
-    // manipulate/step the pointers themselves, rather than repeatedly
-    // .offset-ing.
-    for start in (0..len).step_by(insertion) {
-        // start <= i < len;
-        for i in start..cmp::min(start + insertion, len) {
-            // j satisfies: start <= j <= i;
-            let mut j = i as isize;
-            unsafe {
-                // `i` is in bounds.
-                let read_ptr = buf_v.offset(i as isize);
+    // When dropped, copies the range `start..end` into `dst`.
+    struct Guard<T> {
+        start: *mut T,
+        end: *mut T,
+        dst: *mut T,
+    }
 
-                // find where to insert, we need to do strict <,
-                // rather than <=, to maintain stability.
+    impl<T> Drop for Guard<T> {
+        fn drop(&mut self) {
+            let len = (self.end as usize - self.start as usize) / mem::size_of::<T>();
```

Comment:
If `T` is a zero sized type (ZST), this is a division by zero.

Now I know — the old sort code also does this! The division by size of turns out to be unreachable for ZSTs in that implementation. We can't require much of anything in this case, but maybe a test that ensures that it finishes without panic.